core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala

```diff
@@ -1040,21 +1040,36 @@ class GroupCoordinator(val brokerId: Int,
 
     group.currentState match {
       case Stable =>
-        info(s"Static member joins during Stable stage will not trigger rebalance.")
-        group.maybeInvokeJoinCallback(member, JoinGroupResult(
-          members = List.empty,
-          memberId = newMemberId,
-          generationId = group.generationId,
-          protocolType = group.protocolType,
-          protocolName = group.protocolName,
-          // We want to avoid current leader performing trivial assignment while the group
-          // is in stable stage, because the new assignment in leader's next sync call
-          // won't be broadcast by a stable group. This could be guaranteed by
-          // always returning the old leader id so that the current leader won't assume itself
-          // as a leader based on the returned message, since the new member.id won't match
-          // returned leader id, therefore no assignment will be performed.
-          leaderId = currentLeader,
-          error = Errors.NONE))
+        // check if group's selectedProtocol of next generation will change, if not, simply store group to persist the
+        // updated static member, if yes, rebalance should be triggered to let the group's assignment and selectProtocol consistent
+        val selectedProtocolOfNextGeneration = group.selectProtocol
+        if (group.protocolName.contains(selectedProtocolOfNextGeneration)) {
+          info(s"Static member which joins during Stable stage and doesn't affect selectProtocol will not trigger rebalance.")
+          val groupAssignment: Map[String, Array[Byte]] = group.allMemberMetadata.map(member => member.memberId -> member.assignment).toMap
+          groupManager.storeGroup(group, groupAssignment, error => {
+            group.inLock {
+              if (error != Errors.NONE) {
```

Comment:
Should we reply join failure if the persistence of the group metadata failed?

Reply:
Indeed, here need to be revised, will update~

Reply:
Hi @feyman2016 , I just noticed this PR. Thanks for picking this up!

I'm wondering something along similar lines; should we move the whole join callback inside this block? It seems like that's the way we can delay the join group response until after the update is actually persisted.

Reply:
Thanks for the review, @vvcephei . You're right for the current commit, but as @abbccdda mentioned, the current commit does have issues, I'm revising it to `reset` the `static member` in the group and return the error if `storeGroup` encountered any errors. Code is ready, but met some local issues when writing unittests, will update once it's ready.

Reply:
Revised the persistence failure handling logic, now it will revert the member id update in the groupMetaData if any persistence error encountered and call the responseCallback with the returned error.